Applications/Spire/Source/KeyBindings/KeyBindingsTableView.cpp

```diff
@@ -345,35 +357,20 @@ void KeyBindingsTableView::on_column_selection_changed(
   }
 }
 
-void KeyBindingsTableView::on_data_changed(const QModelIndex& index) {
-  if(index.row() == m_table->model()->rowCount() - 1) {
-    add_delete_button(index.row());
-  }
-}
-
 void KeyBindingsTableView::on_delete_button_clicked(int index) {
   auto table_model = static_cast<KeyBindingsTableModel*>(m_table->model());
   table_model->reset_row_highlight();
   m_table->model()->removeRow(index);
 }
 
-void KeyBindingsTableView::on_editor_key(Qt::Key key) {
-  if(key == Qt::Key_Tab || key == Qt::Key_Backtab) {
-    auto new_index = [&] {
-      if(key == Qt::Key_Tab) {
-        return get_next_editable_index(m_table->currentIndex());
-      }
-      return get_previous_editable_index(m_table->currentIndex());
-    }();
-    m_is_editing_cell = false;
-    m_table->setCurrentIndex(new_index);
-    auto table_model = static_cast<KeyBindingsTableModel*>(m_table->model());
-    table_model->set_focus_highlight(new_index);
-    if(new_index.flags().testFlag(Qt::ItemIsEditable)) {
-      m_is_editing_cell = true;
-      m_table->edit(new_index);
-    }
-    update();
+void KeyBindingsTableView::on_editor_closed(const QWidget* editor,
+    QAbstractItemDelegate::EndEditHint hint) {
+  if(hint == QAbstractItemDelegate::EditNextItem) {
+    auto next = get_next_editable_index(m_table->currentIndex());
+    edit_cell(next);
```

Comment:
Move `edit_cell` out. Use conditional lambda to initialize next/prev if necessary.